pre_commit/languages/download.py

```diff
@@ -0,0 +1,267 @@
+from __future__ import annotations
+
+import contextlib
+import hashlib
+import os.path
+import platform
+import stat
+from base64 import standard_b64decode as b64decode
+from base64 import standard_b64encode as b64encode
+from dataclasses import dataclass
+from os import chmod
+from pathlib import Path
+from pathlib import PurePath
+from types import MappingProxyType
+from typing import Final
+from typing import Generator
+from typing import Sequence
+from urllib.parse import urlparse
+from urllib.request import urlopen
+
+from pre_commit import lang_base
+from pre_commit.envcontext import envcontext
+from pre_commit.envcontext import PatchesT
+from pre_commit.envcontext import Var
+from pre_commit.prefix import Prefix
+
+ENVIRONMENT_DIR = 'download'
+get_default_version = lang_base.basic_get_default_version
+run_hook = lang_base.basic_run_hook
+IO_BUF_SIZE = 4096
+
+
+def get_env_patch(target_dir: str) -> PatchesT:
+    return (
+        ('PATH', (target_dir, os.pathsep, Var('PATH'))),
+    )
+
+
+@contextlib.contextmanager
+def in_env(prefix: Prefix, version: str) -> Generator[None, None, None]:
+    envdir = lang_base.environment_dir(prefix, ENVIRONMENT_DIR, version)
+    with envcontext(get_env_patch(envdir)):
+        yield
+
+
+class Platform:
+    OS: Final[MappingProxyType[str, str]] = MappingProxyType({
+        'Linux': 'linux',
+        'Darwin': 'darwin',
+        'Windows': 'windows',
+        'DragonFly': 'dragonfly',
+        'FreeBSD': 'freebsd',
+    })
+    CPU: Final[MappingProxyType[str, str]] = MappingProxyType({
+        'aarch64': 'arm64',
+        'aarch64_be': 'arm64be',
+        'arm': 'arm',
+        'i386': '386',
+        'i686': '386',
+        'x86_64': 'amd64',
+        'AMD64': 'amd64',
+        'ppc': 'ppc',
+        'ppc64': 'ppc64',
+        'ppc64le': 'ppc64le',
+    })
+
+    def __init__(self, value: str) -> None:
+        self.__value = value
+        os, cpu = self.parts
+        if os not in Platform.OS.values():
+            raise ValueError(f'invalid operating system `{os}`')
+        if cpu not in Platform.CPU.values():
+            raise ValueError(f'invalid CPU `{cpu}`')
```

Comment:
You can use `self` here, which can help if the name of the class ever changes.

You may want to add what the valid values are to the `ValueError`?

```suggestion
        if os not in self.OS.values():
            raise ValueError(f'invalid operating system `{os}`, valid values are: {','.join(self.OS.values())}')
        if cpu not in self.CPU.values():
            raise ValueError(f'invalid CPU `{cpu}`, valid values are: {','.join(self.CPU.values())}')
```

Reply:
Done